File: host/src/machineGrant.ts
import {
  createPrivateKey,
  createPublicKey,
  randomUUID,
  sign,
  verify,
} from "node:crypto";
import type { PairingDevicePublicIdentity } from "./pairing.js";
import type { MachineIdentity } from "./machineIdentity.js";

export interface MachineGrant {
  version: 1;
  grantId: string;
  machine: {
    id: string;
    signingPublicKey: string;
    keyAgreementPublicKey: string;
  };
  device: {
    id: string;
    signingPublicKey: string;
    keyAgreementPublicKey: string;
  };
  role: "owner";
  issuedAt: string;
  signature: string;
}

export type UnsignedMachineGrant = Omit<MachineGrant, "signature">;

function field(value: string): Buffer {
  return Buffer.from(value, "utf8");
}

export function machineGrantMessage(grant: UnsignedMachineGrant): Buffer {
  return Buffer.concat([
    Buffer.from("piremote-machine-grant-v1\0", "utf8"),
    field(grant.grantId), Buffer.from([0]),
    field(grant.machine.id), Buffer.from([0]),
    field(grant.machine.signingPublicKey), Buffer.from([0]),
    field(grant.machine.keyAgreementPublicKey), Buffer.from([0]),
    field(grant.device.id), Buffer.from([0]),
    field(grant.device.signingPublicKey), Buffer.from([0]),
    field(grant.device.keyAgreementPublicKey), Buffer.from([0]),
    field(grant.role), Buffer.from([0]),
    field(grant.issuedAt),
  ]);
}

export function issueMachineGrant(
  machine: MachineIdentity,
  device: PairingDevicePublicIdentity,
  issuedAt: string,
  grantId = "grant_" + randomUUID().replaceAll("-", ""),
): MachineGrant {
  const unsigned: UnsignedMachineGrant = {
    version: 1,
    grantId,
    machine: {
      id: machine.id,
      signingPublicKey: machine.signingPrivateKey.x,
      keyAgreementPublicKey: machine.keyAgreementPrivateKey.x,
    },
    device: {
      id: device.id,
      signingPublicKey: device.signingPublicKey,
      keyAgreementPublicKey: device.keyAgreementPublicKey,
    },
    role: "owner",
    issuedAt,
  };

  const key = machine.signingPrivateKey;
  const privateKey = createPrivateKey({
    key: {
      kty: key.kty,
      crv: key.crv,
      x: key.x,
      d: key.d,
    },
    format: "jwk",
  });

  return {
    ...unsigned,
    signature: sign(null, machineGrantMessage(unsigned), privateKey).toString("base64url"),
  };
}

export function verifyMachineGrant(grant: MachineGrant): boolean {
  if (grant.version !== 1
    || !grant.grantId.startsWith("grant_")
    || !grant.machine.id.startsWith("machine_")
    || !grant.device.id.startsWith("device_")
    || grant.role !== "owner") {
    return false;
  }

  try {
    const publicKey = createPublicKey({
      key: {
        kty: "OKP",
        crv: "Ed25519",
        x: grant.machine.signingPublicKey,
      },
      format: "jwk",
    });

    const { signature, ...unsigned } = grant;
    return verify(
      null,
      machineGrantMessage(unsigned),
      publicKey,
      Buffer.from(signature, "base64url"),
    );
  } catch {
    return false;
  }
}

File: ios/PiRemoteCore/Sources/PiRemoteCore/MachineGrant.swift
import CryptoKit
import Foundation

public struct MachineGrant: Codable, Hashable, Sendable {
    public struct Machine: Codable, Hashable, Sendable {
        public let id: String
        public let signingPublicKey: String
        public let keyAgreementPublicKey: String

        public init(
            id: String,
            signingPublicKey: String,
            keyAgreementPublicKey: String
        ) {
            self.id = id
            self.signingPublicKey = signingPublicKey
            self.keyAgreementPublicKey = keyAgreementPublicKey
        }
    }

    public struct Device: Codable, Hashable, Sendable {
        public let id: String
        public let signingPublicKey: String
        public let keyAgreementPublicKey: String

        public init(
            id: String,
            signingPublicKey: String,
            keyAgreementPublicKey: String
        ) {
            self.id = id
            self.signingPublicKey = signingPublicKey
            self.keyAgreementPublicKey = keyAgreementPublicKey
        }
    }

    public let version: Int
    public let grantId: String
    public let machine: Machine
    public let device: Device
    public let role: String
    public let issuedAt: String
    public let signature: String

    public init(
        version: Int = 1,
        grantId: String,
        machine: Machine,
        device: Device,
        role: String = "owner",
        issuedAt: String,
        signature: String
    ) {
        self.version = version
        self.grantId = grantId
        self.machine = machine
        self.device = device
        self.role = role
        self.issuedAt = issuedAt
        self.signature = signature
    }
}

public enum MachineGrantCrypto {
    public static func message(for grant: MachineGrant) -> Data {
        var message = Data("piremote-machine-grant-v1\0".utf8)
        append(grant.grantId, to: &message)
        append(grant.machine.id, to: &message)
        append(grant.machine.signingPublicKey, to: &message)
        append(grant.machine.keyAgreementPublicKey, to: &message)
        append(grant.device.id, to: &message)
        append(grant.device.signingPublicKey, to: &message)
        append(grant.device.keyAgreementPublicKey, to: &message)
        append(grant.role, to: &message)
        message.append(Data(grant.issuedAt.utf8))
        return message
    }

    public static func isValid(_ grant: MachineGrant) -> Bool {
        guard grant.version == 1,
              grant.grantId.hasPrefix("grant_"),
              grant.machine.id.hasPrefix("machine_"),
              grant.device.id.hasPrefix("device_"),
              grant.role == "owner",
              let publicKeyRaw = Data(base64URLEncoded: grant.machine.signingPublicKey),
              let signature = Data(base64URLEncoded: grant.signature),
              let publicKey = try? Curve25519.Signing.PublicKey(
                rawRepresentation: publicKeyRaw
              )
        else {
            return false
        }

        return publicKey.isValidSignature(signature, for: message(for: grant))
    }

    private static func append(_ value: String, to data: inout Data) {
        data.append(Data(value.utf8))
        data.append(0)
    }
}
